src/libraries/System.Linq/src/System/Linq/Cast.cs

```diff
@@ -104,47 +106,6 @@ public override void Dispose()
 
                 base.Dispose();
             }
-
-            public int Count => _source.Count;
-
-            public bool IsReadOnly => true;
-
-            public int IndexOf(TResult item) => _source.IndexOf(item);
-
-            public bool Contains(TResult item) => _source.Contains(item);
-
-            public void CopyTo(TResult[] array, int arrayIndex)
```

Comment:
It's a shame that we are giving up on potential optimizations just because MD arrays fall out of the usual ICollection mold. Couldn't we just extend our type tests so that MD-arrays are excluded? 

Calling `Cast` on non-generic collections is pretty much its canonical use case so it would be nice if we could improve performance.

Reply:
>  Couldn't we just extend our type tests so that MD-arrays are excluded?

We could, but we'd still blow up for any IList that wrapped a multidimensional array and delegated to it. Once there's one prominent thing that violates the abstraction, I'm a little nervous trying to block just that one thing.